# Task 1 - DataFrame manipulation

Using states_covid.csv, we are going to read the data in as a DataFrame to manipulate, subset, and filter in various ways.

## Task 1.1

Read in states_covid.csv with date as a "date" dtype, and only columns consisting of the hospitalization (4 col), ICU (2 col), and Ventilators (2 col)

In [45]:
import os
import numpy as np
import pandas as pd

In [67]:
cd Desktop/DrAllen_Biol792

[Errno 2] No such file or directory: 'Desktop/DrAllen_Biol792'
/Users/erickarkay/Desktop/Julies_Class


In [66]:
ls

Bloom_etal_2018_Reduced_Dataset.csv  Kay_E_HW4.md
EK_HW4.webarchive                    Kay_E_HW4.tex
HW_3.txt                             ggplot/
Julie_HW_wk1/                        logfiles/
Kay_E_HW1.docx                       scripts/
Kay_E_HW1.txt                        shell key points .pdf
Kay_E_HW2.txt                        shell-lesson-data/
Kay_E_HW3.ipynb                      shell-lesson-data.zip
Kay_E_HW4.ipynb                      states_covid.csv


In [59]:
state_covid_df = pd.read_csv("states_covid.csv")

In [60]:
state_covid_df.head()

date state    death  deathConfirmed  deathIncrease  deathProbable  \
0  2021-02-23    AK    290.0             NaN              0            NaN   
1  2021-02-23    AL   9660.0          7575.0             68         2085.0   
2  2021-02-23    AR   5377.0          4321.0             14         1056.0   
3  2021-02-23    AS      0.0             NaN              0            NaN   
4  2021-02-23    AZ  15650.0         13821.0            148         1829.0   

   hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
0        1260.0                  1260.0                   38.0   
1       45250.0                 45250.0                  762.0   
2       14617.0                 14617.0                  545.0   
3           NaN                     NaN                    NaN   
4       57072.0                 57072.0                 1515.0   

   hospitalizedIncrease  ...  totalTestResults  totalTestResultsIncrease  \
0                     9  ...         1653425.0                      4640   
1                   122  ...         2265086.0                      4825   
2                    47  ...         2609837.0                      4779   
3                     0  ...            2140.0                         0   
4                    78  ...         7478323.0                     19439   

   totalTestsAntibody  totalTestsAntigen  totalTestsPeopleAntibody  \
0                 NaN                NaN                       NaN   
1                 NaN                NaN                  115256.0   
2                 NaN                NaN                       NaN   
3                 NaN                NaN                       NaN   
4            435091.0                NaN                       NaN   

   totalTestsPeopleAntigen  totalTestsPeopleViral  \
0                      NaN                    NaN   
1                      NaN              2265086.0   
2                 436309.0                    NaN   
3                      NaN                    NaN   
4                      NaN              3709365.0   

   totalTestsPeopleViralIncrease  totalTestsViral  totalTestsViralIncrease  
0                              0        1653425.0                     4640  
1                           4825              NaN                        0  
2                              0        2609837.0                     4779  
3                              0           2140.0                        0  
4                           6212        7478323.0                    19439  

[5 rows x 41 columns]

In [61]:
new_covid_df = state_covid_df[['date', 'state','hospitalized', 'hospitalizedCumulative', 'hospitalizedCurrently', 'hospitalizedIncrease','inIcuCurrently','inIcuCumulative','onVentilatorCurrently', 'onVentilatorCumulative']]

In [62]:
new_covid_df.head()

date state  hospitalized  hospitalizedCumulative  \
0  2021-02-23    AK        1260.0                  1260.0   
1  2021-02-23    AL       45250.0                 45250.0   
2  2021-02-23    AR       14617.0                 14617.0   
3  2021-02-23    AS           NaN                     NaN   
4  2021-02-23    AZ       57072.0                 57072.0   

   hospitalizedCurrently  hospitalizedIncrease  inIcuCurrently  \
0                   38.0                     9             NaN   
1                  762.0                   122             NaN   
2                  545.0                    47           204.0   
3                    NaN                     0             NaN   
4                 1515.0                    78           447.0   

   inIcuCumulative  onVentilatorCurrently  onVentilatorCumulative  
0              NaN                    5.0                     NaN  
1           2632.0                    NaN                  1497.0  
2              NaN                   99.0                  1505.0  
3              NaN                    NaN                     NaN  
4              NaN                  266.0                     NaN

In [41]:
new_covid_df['date'] = pd.to_datetime(new_covid_df['date'],yearfirst=True) 
new_covid_df.dtypes

<ipython-input-41-f7119624ae56>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_covid_df['date'] = pd.to_datetime(new_covid_df['date'],yearfirst=True)


date                      datetime64[ns]
state                                |S2
hospitalized                     float64
hospitalizedCumulative           float64
hospitalizedCurrently            float64
hospitalizedIncrease               int64
inIcuCurrently                   float64
inIcuCumulative                  float64
onVentilatorCurrently            float64
onVentilatorCumulative           float64
dtype: object

In [11]:
new_covid_df['state'] = new_covid_df['state'].astype("|S")

<ipython-input-11-ed37d5e594ea>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_covid_df['state'] = new_covid_df['state'].astype("|S")


In [12]:
new_covid_df.dtypes

date                      datetime64[ns]
state                                |S2
hospitalized                     float64
hospitalizedCumulative           float64
hospitalizedCurrently            float64
hospitalizedIncrease               int64
inIcuCurrently                   float64
inIcuCumulative                  float64
onVentilatorCurrently            float64
onVentilatorCumulative           float64
dtype: object

In [13]:
display(new_covid_df)

date  state  hospitalized  hospitalizedCumulative  \
0     2021-02-23  b'AK'        1260.0                  1260.0   
1     2021-02-23  b'AL'       45250.0                 45250.0   
2     2021-02-23  b'AR'       14617.0                 14617.0   
3     2021-02-23  b'AS'           NaN                     NaN   
4     2021-02-23  b'AZ'       57072.0                 57072.0   
...          ...    ...           ...                     ...   
20103 2020-01-17  b'WA'           NaN                     NaN   
20104 2020-01-16  b'WA'           NaN                     NaN   
20105 2020-01-15  b'WA'           NaN                     NaN   
20106 2020-01-14  b'WA'           NaN                     NaN   
20107 2020-01-13  b'WA'           NaN                     NaN   

       hospitalizedCurrently  hospitalizedIncrease  inIcuCurrently  \
0                       38.0                     9             NaN   
1                      762.0                   122             NaN   
2                      545.0                    47           204.0   
3                        NaN                     0             NaN   
4                     1515.0                    78           447.0   
...                      ...                   ...             ...   
20103                    NaN                     0             NaN   
20104                    NaN                     0             NaN   
20105                    NaN                     0             NaN   
20106                    NaN                     0             NaN   
20107                    NaN                     0             NaN   

       inIcuCumulative  onVentilatorCurrently  onVentilatorCumulative  
0                  NaN                    5.0                     NaN  
1               2632.0                    NaN                  1497.0  
2                  NaN                   99.0                  1505.0  
3                  NaN                    NaN                     NaN  
4                  NaN                  266.0                     NaN  
...                ...                    ...                     ...  
20103              NaN                    NaN                     NaN  
20104              NaN                    NaN                     NaN  
20105              NaN                    NaN                     NaN  
20106              NaN                    NaN                     NaN  
20107              NaN                    NaN                     NaN  

[20108 rows x 10 columns]

## Task 1.2

For each of the following catergories: currently hospitalized, currently in the ICU, and currently on ventilation...
Find the 5 states with the greatest numbers in each catergory and list them in order.
hint: sort_values, unique


In [14]:
current_hosp = new_covid_df.groupby('state').max('hospitalizedCurrently')

In [15]:
current_hosp.head(10)

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
b'AK'        1260.0                  1260.0                  166.0   
b'AL'       45250.0                 45250.0                 3088.0   
b'AR'       14617.0                 14617.0                 1371.0   
b'AS'           NaN                     NaN                    NaN   
b'AZ'       57072.0                 57072.0                 5082.0   
b'CA'           NaN                     NaN                22851.0   
b'CO'       23293.0                 23293.0                 1995.0   
b'CT'       13793.0                 13793.0                 1972.0   
b'DC'           NaN                     NaN                  447.0   
b'DE'           NaN                     NaN                  474.0   

       hospitalizedIncrease  inIcuCurrently  inIcuCumulative  \
state                                                          
b'AK'                    34             NaN              NaN   
b'AL'                  3656             NaN           2632.0   
b'AR'                   229           460.0             43.0   
b'AS'                     0             NaN              NaN   
b'AZ'                  5173          1183.0              NaN   
b'CA'                     0          4971.0              NaN   
b'CO'                   885             NaN              NaN   
b'CT'                  7758             NaN              NaN   
b'DC'                     0           130.0              NaN   
b'DE'                     0            68.0              NaN   

       onVentilatorCurrently  onVentilatorCumulative  
state                                                 
b'AK'                   27.0                     NaN  
b'AL'                    NaN                  1497.0  
b'AR'                  319.0                  1505.0  
b'AS'                    NaN                     NaN  
b'AZ'                  821.0                     NaN  
b'CA'                    NaN                     NaN  
b'CO'                    NaN                     NaN  
b'CT'                    NaN                     NaN  
b'DC'                  200.0                     NaN  
b'DE'                    NaN                     NaN

In [16]:
current_hosp.sort_values(['hospitalizedCurrently'], ascending = (False)) 

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
b'CA'           NaN                     NaN                22851.0   
b'NY'       89995.0                 89995.0                18825.0   
b'TX'           NaN                     NaN                14218.0   
b'FL'       79410.0                 79410.0                 9520.0   
b'NJ'       63190.0                 63190.0                 8270.0   
b'PA'        1145.0                  1145.0                 6346.0   
b'IL'           NaN                     NaN                 6175.0   
b'GA'       55167.0                 55167.0                 6108.0   
b'OH'       49651.0                 49651.0                 5296.0   
b'AZ'       57072.0                 57072.0                 5082.0   
b'MI'           NaN                     NaN                 4326.0   
b'NC'           NaN                     NaN                 3990.0   
b'MA'       19176.0                 19176.0                 3977.0   
b'TN'       18363.0                 18363.0                 3554.0   
b'IN'       42557.0                 42557.0                 3460.0   
b'VA'       23698.0                 23698.0                 3209.0   
b'AL'       45250.0                 45250.0                 3088.0   
b'MO'           NaN                     NaN                 2862.0   
b'SC'       20024.0                 20024.0                 2466.0   
b'WI'       25838.0                 25838.0                 2278.0   
b'LA'           NaN                     NaN                 2134.0   
b'NV'           NaN                     NaN                 2025.0   
b'CO'       23293.0                 23293.0                 1995.0   
b'OK'       23700.0                 23700.0                 1994.0   
b'CT'       13793.0                 13793.0                 1972.0   
b'MD'       34577.0                 34577.0                 1957.0   
b'MN'       25528.0                 25528.0                 1840.0   
b'KY'       18768.0                 18768.0                 1817.0   
b'IA'         193.0                   193.0                 1527.0   
b'MS'        9012.0                  9012.0                 1518.0   
b'AR'       14617.0                 14617.0                 1371.0   
b'WA'       19110.0                 19110.0                 1280.0   
b'KS'        9103.0                  9103.0                 1196.0   
b'NE'        6048.0                  6048.0                  987.0   
b'NM'       12989.0                 12989.0                  947.0   
b'WV'         164.0                   164.0                  818.0   
b'PR'           NaN                     NaN                  657.0   
b'OR'        8457.0                  8457.0                  642.0   
b'UT'       14520.0                 14520.0                  638.0   
b'SD'        6548.0                  6548.0                  607.0   
b'MT'        4554.0                  4554.0                  506.0   
b'RI'        8870.0                  8870.0                  479.0   
b'ID'        7007.0                  7007.0                  477.0   
b'DE'           NaN                     NaN                  474.0   
b'DC'           NaN                     NaN                  447.0   
b'ND'        3860.0                  3860.0                  425.0   
b'NH'        1095.0                  1095.0                  335.0   
b'HI'        2210.0                  2210.0                  297.0   
b'WY'        1373.0                  1373.0                  247.0   
b'ME'        1511.0                  1511.0                  207.0   
b'AK'        1260.0                  1260.0                  166.0   
b'GU'           NaN                     NaN                  102.0   
b'VT'          50.0                    50.0                   77.0   
b'VI'           NaN                     NaN                    3.0   
b'AS'           NaN                     NaN                    NaN   
b'MP'           4

Currently hosptiilizied ANSWER: CA, NY, TX, FL, NJ

In [17]:
current_ICU = new_covid_df.groupby('state').max('inIcuCurrently')

In [18]:
current_ICU.head(10)

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
b'AK'        1260.0                  1260.0                  166.0   
b'AL'       45250.0                 45250.0                 3088.0   
b'AR'       14617.0                 14617.0                 1371.0   
b'AS'           NaN                     NaN                    NaN   
b'AZ'       57072.0                 57072.0                 5082.0   
b'CA'           NaN                     NaN                22851.0   
b'CO'       23293.0                 23293.0                 1995.0   
b'CT'       13793.0                 13793.0                 1972.0   
b'DC'           NaN                     NaN                  447.0   
b'DE'           NaN                     NaN                  474.0   

       hospitalizedIncrease  inIcuCurrently  inIcuCumulative  \
state                                                          
b'AK'                    34             NaN              NaN   
b'AL'                  3656             NaN           2632.0   
b'AR'                   229           460.0             43.0   
b'AS'                     0             NaN              NaN   
b'AZ'                  5173          1183.0              NaN   
b'CA'                     0          4971.0              NaN   
b'CO'                   885             NaN              NaN   
b'CT'                  7758             NaN              NaN   
b'DC'                     0           130.0              NaN   
b'DE'                     0            68.0              NaN   

       onVentilatorCurrently  onVentilatorCumulative  
state                                                 
b'AK'                   27.0                     NaN  
b'AL'                    NaN                  1497.0  
b'AR'                  319.0                  1505.0  
b'AS'                    NaN                     NaN  
b'AZ'                  821.0                     NaN  
b'CA'                    NaN                     NaN  
b'CO'                    NaN                     NaN  
b'CT'                    NaN                     NaN  
b'DC'                  200.0                     NaN  
b'DE'                    NaN                     NaN

In [19]:
current_ICU.sort_values(['inIcuCurrently'], ascending = (False)) 

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
b'NY'       89995.0                 89995.0                18825.0   
b'CA'           NaN                     NaN                22851.0   
b'TX'           NaN                     NaN                14218.0   
b'NJ'       63190.0                 63190.0                 8270.0   
b'MI'           NaN                     NaN                 4326.0   
b'OH'       49651.0                 49651.0                 5296.0   
b'IL'           NaN                     NaN                 6175.0   
b'PA'        1145.0                  1145.0                 6346.0   
b'AZ'       57072.0                 57072.0                 5082.0   
b'MA'       19176.0                 19176.0                 3977.0   
b'IN'       42557.0                 42557.0                 3460.0   
b'NC'           NaN                     NaN                 3990.0   
b'TN'       18363.0                 18363.0                 3554.0   
b'MO'           NaN                     NaN                 2862.0   
b'MD'       34577.0                 34577.0                 1957.0   
b'VA'       23698.0                 23698.0                 3209.0   
b'KY'       18768.0                 18768.0                 1817.0   
b'OK'       23700.0                 23700.0                 1994.0   
b'SC'       20024.0                 20024.0                 2466.0   
b'NV'           NaN                     NaN                 2025.0   
b'AR'       14617.0                 14617.0                 1371.0   
b'WI'       25838.0                 25838.0                 2278.0   
b'MN'       25528.0                 25528.0                 1840.0   
b'MS'        9012.0                  9012.0                 1518.0   
b'KS'        9103.0                  9103.0                 1196.0   
b'WA'       19110.0                 19110.0                 1280.0   
b'IA'         193.0                   193.0                 1527.0   
b'UT'       14520.0                 14520.0                  638.0   
b'WV'         164.0                   164.0                  818.0   
b'OR'        8457.0                  8457.0                  642.0   
b'DC'           NaN                     NaN                  447.0   
b'ID'        7007.0                  7007.0                  477.0   
b'SD'        6548.0                  6548.0                  607.0   
b'PR'           NaN                     NaN                  657.0   
b'MT'        4554.0                  4554.0                  506.0   
b'RI'        8870.0                  8870.0                  479.0   
b'NE'        6048.0                  6048.0                  987.0   
b'ME'        1511.0                  1511.0                  207.0   
b'DE'           NaN                     NaN                  474.0   
b'HI'        2210.0                  2210.0                  297.0   
b'ND'        3860.0                  3860.0                  425.0   
b'GU'           NaN                     NaN                  102.0   
b'VT'          50.0                    50.0                   77.0   
b'AK'        1260.0                  1260.0                  166.0   
b'AL'       45250.0                 45250.0                 3088.0   
b'AS'           NaN                     NaN                    NaN   
b'CO'       23293.0                 23293.0                 1995.0   
b'CT'       13793.0                 13793.0                 1972.0   
b'FL'       79410.0                 79410.0                 9520.0   
b'GA'       55167.0                 55167.0                 6108.0   
b'LA'           NaN                     NaN                 2134.0   
b'MP'           4.0                     4.0                    NaN   
b'NH'        1095.0                  1095.0                  335.0   
b'NM'       12989.0                 12989.0                  947.0   
b'VI'           NaN                     NaN                    3.0   
b'WY'        1373

CURRENTLY IN ICU ANSWER: NY, CA, TX, NY, NJ, MI

In [20]:
current_vent = new_covid_df.groupby('state').max('onVentilatorCurrently')

In [21]:
current_vent.head(10)

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
b'AK'        1260.0                  1260.0                  166.0   
b'AL'       45250.0                 45250.0                 3088.0   
b'AR'       14617.0                 14617.0                 1371.0   
b'AS'           NaN                     NaN                    NaN   
b'AZ'       57072.0                 57072.0                 5082.0   
b'CA'           NaN                     NaN                22851.0   
b'CO'       23293.0                 23293.0                 1995.0   
b'CT'       13793.0                 13793.0                 1972.0   
b'DC'           NaN                     NaN                  447.0   
b'DE'           NaN                     NaN                  474.0   

       hospitalizedIncrease  inIcuCurrently  inIcuCumulative  \
state                                                          
b'AK'                    34             NaN              NaN   
b'AL'                  3656             NaN           2632.0   
b'AR'                   229           460.0             43.0   
b'AS'                     0             NaN              NaN   
b'AZ'                  5173          1183.0              NaN   
b'CA'                     0          4971.0              NaN   
b'CO'                   885             NaN              NaN   
b'CT'                  7758             NaN              NaN   
b'DC'                     0           130.0              NaN   
b'DE'                     0            68.0              NaN   

       onVentilatorCurrently  onVentilatorCumulative  
state                                                 
b'AK'                   27.0                     NaN  
b'AL'                    NaN                  1497.0  
b'AR'                  319.0                  1505.0  
b'AS'                    NaN                     NaN  
b'AZ'                  821.0                     NaN  
b'CA'                    NaN                     NaN  
b'CO'                    NaN                     NaN  
b'CT'                    NaN                     NaN  
b'DC'                  200.0                     NaN  
b'DE'                    NaN                     NaN

In [22]:
current_vent.sort_values(['onVentilatorCurrently'], ascending = (False))

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
b'NY'       89995.0                 89995.0                18825.0   
b'NJ'       63190.0                 63190.0                 8270.0   
b'MI'           NaN                     NaN                 4326.0   
b'OH'       49651.0                 49651.0                 5296.0   
b'AZ'       57072.0                 57072.0                 5082.0   
b'IL'           NaN                     NaN                 6175.0   
b'PA'        1145.0                  1145.0                 6346.0   
b'LA'           NaN                     NaN                 2134.0   
b'TN'       18363.0                 18363.0                 3554.0   
b'IN'       42557.0                 42557.0                 3460.0   
b'MO'           NaN                     NaN                 2862.0   
b'VA'       23698.0                 23698.0                 3209.0   
b'WI'       25838.0                 25838.0                 2278.0   
b'AR'       14617.0                 14617.0                 1371.0   
b'SC'       20024.0                 20024.0                 2466.0   
b'NV'           NaN                     NaN                 2025.0   
b'MA'       19176.0                 19176.0                 3977.0   
b'KY'       18768.0                 18768.0                 1817.0   
b'MS'        9012.0                  9012.0                 1518.0   
b'DC'           NaN                     NaN                  447.0   
b'IA'         193.0                   193.0                 1527.0   
b'WA'       19110.0                 19110.0                 1280.0   
b'KS'        9103.0                  9103.0                 1196.0   
b'PR'           NaN                     NaN                  657.0   
b'WV'         164.0                   164.0                  818.0   
b'OR'        8457.0                  8457.0                  642.0   
b'SD'        6548.0                  6548.0                  607.0   
b'RI'        8870.0                  8870.0                  479.0   
b'NE'        6048.0                  6048.0                  987.0   
b'MT'        4554.0                  4554.0                  506.0   
b'HI'        2210.0                  2210.0                  297.0   
b'ME'        1511.0                  1511.0                  207.0   
b'AK'        1260.0                  1260.0                  166.0   
b'NM'       12989.0                 12989.0                  947.0   
b'GU'           NaN                     NaN                  102.0   
b'VT'          50.0                    50.0                   77.0   
b'AL'       45250.0                 45250.0                 3088.0   
b'AS'           NaN                     NaN                    NaN   
b'CA'           NaN                     NaN                22851.0   
b'CO'       23293.0                 23293.0                 1995.0   
b'CT'       13793.0                 13793.0                 1972.0   
b'DE'           NaN                     NaN                  474.0   
b'FL'       79410.0                 79410.0                 9520.0   
b'GA'       55167.0                 55167.0                 6108.0   
b'ID'        7007.0                  7007.0                  477.0   
b'MD'       34577.0                 34577.0                 1957.0   
b'MN'       25528.0                 25528.0                 1840.0   
b'MP'           4.0                     4.0                    NaN   
b'NC'           NaN                     NaN                 3990.0   
b'ND'        3860.0                  3860.0                  425.0   
b'NH'        1095.0                  1095.0                  335.0   
b'OK'       23700.0                 23700.0                 1994.0   
b'TX'           NaN                     NaN                14218.0   
b'UT'       14520.0                 14520.0                  638.0   
b'VI'           NaN                     NaN                    3.0   
b'WY'        1373

Currently on ventilator answer: NY, NJ, MI, OH, AZ

### Task 1.3

Find the date in which each state crossed 1000 cumulative hospitilized covid patients

In [28]:
cum_hosp_df = new_covid_df[(new_covid_df.hospitalizedCumulative > 1000)]

### I found multiple ways to group DF by state and minimum hosptilized cumalitive, but NONE of them included the date. 

In [31]:
cum_hosp_df.groupby('state').min('hospitalizedCumulative')

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
b'AK'        1004.0                  1004.0                   32.0   
b'AL'        1008.0                  1008.0                  420.0   
b'AR'        1003.0                  1003.0                  206.0   
b'AZ'        1037.0                  1037.0                  468.0   
b'CO'        1079.0                  1079.0                  211.0   
b'CT'        7758.0                  7758.0                   42.0   
b'FL'        1123.0                  1123.0                 2005.0   
b'GA'        1056.0                  1056.0                  783.0   
b'HI'        1001.0                  1001.0                   37.0   
b'ID'        1006.0                  1006.0                  105.0   
b'IN'        4389.0                  4389.0                  595.0   
b'KS'        1011.0                  1011.0                  140.0   
b'KY'        1008.0                  1008.0                  263.0   
b'MA'        1370.0                  1370.0                  290.0   
b'MD'        1059.0                  1059.0                  281.0   
b'ME'        1015.0                  1015.0                   67.0   
b'MN'        1044.0                  1044.0                  227.0   
b'MS'        1001.0                  1001.0                  465.0   
b'MT'        1001.0                  1001.0                   78.0   
b'ND'        1016.0                  1016.0                   31.0   
b'NE'        1029.0                  1029.0                   95.0   
b'NH'        1007.0                  1007.0                  109.0   
b'NJ'       16373.0                 16373.0                  349.0   
b'NM'        1139.0                  1139.0                   59.0   
b'NY'        1531.0                  1531.0                  410.0   
b'OH'        1006.0                  1006.0                  482.0   
b'OK'        1003.0                  1003.0                  136.0   
b'OR'        1006.0                  1006.0                  132.0   
b'PA'        1004.0                  1004.0                    NaN   
b'RI'        1013.0                  1013.0                   55.0   
b'SC'        1110.0                  1110.0                  387.0   
b'SD'        1006.0                  1006.0                   68.0   
b'TN'        1013.0                  1013.0                  463.0   
b'UT'        1012.0                  1012.0                  127.0   
b'VA'        1048.0                  1048.0                  783.0   
b'WA'        3125.0                  3125.0                  342.0   
b'WI'        1049.0                  1049.0                  165.0   
b'WY'        1001.0                  1001.0                   21.0   

       hospitalizedIncrease  inIcuCurrently  inIcuCumulative  \
state                                                          
b'AK'                     0             NaN              NaN   
b'AL'                     0             NaN            335.0   
b'AR'                    -4           160.0              NaN   
b'AZ'                 -2316           114.0              NaN   
b'CO'                   -13             NaN              NaN   
b'CT'                 -4124             NaN              NaN   
b'FL'                     0             NaN              NaN   
b'GA'                     0             NaN           1057.0   
b'HI'                   -14             7.0              NaN   
b'ID'                    -1            25.0            282.0   
b'IN'                   -13           159.0            990.0   
b'KS'                     0             1.0            333.0   
b'KY'                    -5            61.0            514.0   
b'MA'                   -91            47.0              NaN   
b'MD'                     0            68.0              NaN   
b'ME'                    -2            22.0              NaN   
b'MN'                     0            48.0           

In [32]:
cum_hosp_df.groupby('state')['hospitalizedCumulative'].agg([min])

min
state         
b'AK'   1004.0
b'AL'   1008.0
b'AR'   1003.0
b'AZ'   1037.0
b'CO'   1079.0
b'CT'   7758.0
b'FL'   1123.0
b'GA'   1056.0
b'HI'   1001.0
b'ID'   1006.0
b'IN'   4389.0
b'KS'   1011.0
b'KY'   1008.0
b'MA'   1370.0
b'MD'   1059.0
b'ME'   1015.0
b'MN'   1044.0
b'MS'   1001.0
b'MT'   1001.0
b'ND'   1016.0
b'NE'   1029.0
b'NH'   1007.0
b'NJ'  16373.0
b'NM'   1139.0
b'NY'   1531.0
b'OH'   1006.0
b'OK'   1003.0
b'OR'   1006.0
b'PA'   1004.0
b'RI'   1013.0
b'SC'   1110.0
b'SD'   1006.0
b'TN'   1013.0
b'UT'   1012.0
b'VA'   1048.0
b'WA'   3125.0
b'WI'   1049.0
b'WY'   1001.0

In [33]:
cum_hosp_df.groupby('state').agg({'hospitalizedCumulative': ['min']})

hospitalizedCumulative
                         min
state                       
b'AK'                 1004.0
b'AL'                 1008.0
b'AR'                 1003.0
b'AZ'                 1037.0
b'CO'                 1079.0
b'CT'                 7758.0
b'FL'                 1123.0
b'GA'                 1056.0
b'HI'                 1001.0
b'ID'                 1006.0
b'IN'                 4389.0
b'KS'                 1011.0
b'KY'                 1008.0
b'MA'                 1370.0
b'MD'                 1059.0
b'ME'                 1015.0
b'MN'                 1044.0
b'MS'                 1001.0
b'MT'                 1001.0
b'ND'                 1016.0
b'NE'                 1029.0
b'NH'                 1007.0
b'NJ'                16373.0
b'NM'                 1139.0
b'NY'                 1531.0
b'OH'                 1006.0
b'OK'                 1003.0
b'OR'                 1006.0
b'PA'                 1004.0
b'RI'                 1013.0
b'SC'                 1110.0
b'SD'                 1006.0
b'TN'                 1013.0
b'UT'                 1012.0
b'VA'                 1048.0
b'WA'                 3125.0
b'WI'                 1049.0
b'WY'                 1001.0

In [34]:
cum_hosp_df.groupby('state').min('hospitalizedCumulative')

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
b'AK'        1004.0                  1004.0                   32.0   
b'AL'        1008.0                  1008.0                  420.0   
b'AR'        1003.0                  1003.0                  206.0   
b'AZ'        1037.0                  1037.0                  468.0   
b'CO'        1079.0                  1079.0                  211.0   
b'CT'        7758.0                  7758.0                   42.0   
b'FL'        1123.0                  1123.0                 2005.0   
b'GA'        1056.0                  1056.0                  783.0   
b'HI'        1001.0                  1001.0                   37.0   
b'ID'        1006.0                  1006.0                  105.0   
b'IN'        4389.0                  4389.0                  595.0   
b'KS'        1011.0                  1011.0                  140.0   
b'KY'        1008.0                  1008.0                  263.0   
b'MA'        1370.0                  1370.0                  290.0   
b'MD'        1059.0                  1059.0                  281.0   
b'ME'        1015.0                  1015.0                   67.0   
b'MN'        1044.0                  1044.0                  227.0   
b'MS'        1001.0                  1001.0                  465.0   
b'MT'        1001.0                  1001.0                   78.0   
b'ND'        1016.0                  1016.0                   31.0   
b'NE'        1029.0                  1029.0                   95.0   
b'NH'        1007.0                  1007.0                  109.0   
b'NJ'       16373.0                 16373.0                  349.0   
b'NM'        1139.0                  1139.0                   59.0   
b'NY'        1531.0                  1531.0                  410.0   
b'OH'        1006.0                  1006.0                  482.0   
b'OK'        1003.0                  1003.0                  136.0   
b'OR'        1006.0                  1006.0                  132.0   
b'PA'        1004.0                  1004.0                    NaN   
b'RI'        1013.0                  1013.0                   55.0   
b'SC'        1110.0                  1110.0                  387.0   
b'SD'        1006.0                  1006.0                   68.0   
b'TN'        1013.0                  1013.0                  463.0   
b'UT'        1012.0                  1012.0                  127.0   
b'VA'        1048.0                  1048.0                  783.0   
b'WA'        3125.0                  3125.0                  342.0   
b'WI'        1049.0                  1049.0                  165.0   
b'WY'        1001.0                  1001.0                   21.0   

       hospitalizedIncrease  inIcuCurrently  inIcuCumulative  \
state                                                          
b'AK'                     0             NaN              NaN   
b'AL'                     0             NaN            335.0   
b'AR'                    -4           160.0              NaN   
b'AZ'                 -2316           114.0              NaN   
b'CO'                   -13             NaN              NaN   
b'CT'                 -4124             NaN              NaN   
b'FL'                     0             NaN              NaN   
b'GA'                     0             NaN           1057.0   
b'HI'                   -14             7.0              NaN   
b'ID'                    -1            25.0            282.0   
b'IN'                   -13           159.0            990.0   
b'KS'                     0             1.0            333.0   
b'KY'                    -5            61.0            514.0   
b'MA'                   -91            47.0              NaN   
b'MD'                     0            68.0              NaN   
b'ME'                    -2            22.0              NaN   
b'MN'                     0            48.0           

### below includes date column but I do NOT understand how hosptilizedCumalitize is what the 'min' funciton is acting on since i did not specify it. I have spent hours on this and am walking away confused and frustrated. 

In [35]:
cum_hosp_df.groupby('state').agg('min')

date  hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                           
b'AK' 2020-12-29        1004.0                  1004.0                   32.0   
b'AL' 2020-05-01        1008.0                  1008.0                  420.0   
b'AR' 2020-06-15        1003.0                  1003.0                  206.0   
b'AZ' 2020-04-11        1037.0                  1037.0                  468.0   
b'CO' 2020-04-07        1079.0                  1079.0                  211.0   
b'CT' 2020-05-01        7758.0                  7758.0                   42.0   
b'FL' 2020-04-02        1123.0                  1123.0                 2005.0   
b'GA' 2020-04-02        1056.0                  1056.0                  783.0   
b'HI' 2020-10-18        1001.0                  1001.0                   37.0   
b'ID' 2020-08-11        1006.0                  1006.0                  105.0   
b'IN' 2020-05-08        4389.0                  4389.0                  595.0   
b'KS' 2020-06-17        1011.0                  1011.0                  140.0   
b'KY' 2020-04-18        1008.0                  1008.0                  263.0   
b'MA' 2020-04-04        1370.0                  1370.0                  290.0   
b'MD' 2020-04-06        1059.0                  1059.0                  281.0   
b'ME' 2020-12-24        1015.0                  1015.0                   67.0   
b'MN' 2020-04-30        1044.0                  1044.0                  227.0   
b'MS' 2020-04-26        1001.0                  1001.0                  465.0   
b'MT' 2020-10-17        1001.0                  1001.0                   78.0   
b'ND' 2020-10-08        1016.0                  1016.0                   31.0   
b'NE' 2020-06-12        1029.0                  1029.0                   95.0   
b'NH' 2021-01-25        1007.0                  1007.0                  109.0   
b'NJ' 2020-05-26       16373.0                 16373.0                  349.0   
b'NM' 2020-05-20        1139.0                  1139.0                   59.0   
b'NY' 2020-03-21        1531.0                  1531.0                  410.0   
b'OH' 2020-04-04        1006.0                  1006.0                  482.0   
b'OK' 2020-06-03        1003.0                  1003.0                  136.0   
b'OR' 2020-06-25        1006.0                  1006.0                  132.0   
b'PA' 2020-04-04        1004.0                  1004.0                    NaN   
b'RI' 2020-05-03        1013.0                  1013.0                   55.0   
b'SC' 2020-05-01        1110.0                  1110.0                  387.0   
b'SD' 2020-08-29        1006.0                  1006.0                   68.0   
b'TN' 2020-04-29        1013.0                  1013.0                  463.0   
b'UT' 2020-06-13        1012.0                  1012.0                  127.0   
b'VA' 2020-04-16        1048.0                  1048.0                  783.0   
b'WA' 2020-05-21        3125.0                  3125.0                  342.0   
b'WI' 2020-04-14        1049.0                  1049.0                  165.0   
b'WY' 2020-12-18        1001.0                  1001.0                   21.0   

       hospitalizedIncrease  inIcuCurrently  inIcuCumulative  \
state                                                          
b'AK'                     0             NaN              NaN   
b'AL'                     0             NaN            335.0   
b'AR'                    -4           160.0              NaN   
b'AZ'                 -2316           114.0              NaN   
b'CO'                   -13             NaN              NaN   
b'CT'                 -4124             NaN              NaN   
b'FL'                     0             NaN              NaN   
b'GA'                     0             NaN           1057.0   
b'HI'                   -14             7.0              NaN   
b'ID'                    -1            25.0            282.0   
b'I

In [40]:
cum_hosp_df

date  state  hospitalized  hospitalizedCumulative  \
0     2021-02-23  b'AK'        1260.0                  1260.0   
1     2021-02-23  b'AL'       45250.0                 45250.0   
2     2021-02-23  b'AR'       14617.0                 14617.0   
4     2021-02-23  b'AZ'       57072.0                 57072.0   
6     2021-02-23  b'CO'       23293.0                 23293.0   
...          ...    ...           ...                     ...   
18797 2020-03-25  b'NY'        5126.0                  5126.0   
18853 2020-03-24  b'NY'        4041.0                  4041.0   
18909 2020-03-23  b'NY'        3125.0                  3125.0   
18965 2020-03-22  b'NY'        2354.0                  2354.0   
19021 2020-03-21  b'NY'        1531.0                  1531.0   

       hospitalizedCurrently  hospitalizedIncrease  inIcuCurrently  \
0                       38.0                     9             NaN   
1                      762.0                   122             NaN   
2                      545.0                    47           204.0   
4                     1515.0                    78           447.0   
6                      447.0                   110             NaN   
...                      ...                   ...             ...   
18797                 4079.0                  1085             NaN   
18853                 3343.0                   916             NaN   
18909                 2629.0                   771             NaN   
18965                 2043.0                   823             NaN   
19021                 1436.0                  1531             NaN   

       inIcuCumulative  onVentilatorCurrently  onVentilatorCumulative  
0                  NaN                    5.0                     NaN  
1               2632.0                    NaN                  1497.0  
2                  NaN                   99.0                  1505.0  
4                  NaN                  266.0                     NaN  
6                  NaN                    NaN                     NaN  
...                ...                    ...                     ...  
18797              NaN                    NaN                     NaN  
18853              NaN                    NaN                     NaN  
18909              NaN                    NaN                     NaN  
18965              NaN                    NaN                     NaN  
19021              NaN                    NaN                     NaN  

[9303 rows x 10 columns]

### playing with code to figure out why it works how it does

In [36]:
df = new_covid_df[['date', 'state','hospitalizedCurrently', 'hospitalizedIncrease','inIcuCurrently','inIcuCumulative','onVentilatorCurrently', 'onVentilatorCumulative']]

In [39]:
df.groupby('state').agg('min')

date  hospitalizedCurrently  hospitalizedIncrease  inIcuCurrently  \
state                                                                           
b'AK' 2020-03-06                    7.0                     0             NaN   
b'AL' 2020-03-07                   74.0                     0             NaN   
b'AR' 2020-03-06                   13.0                   -18           160.0   
b'AS' 2020-03-16                    NaN                     0             NaN   
b'AZ' 2020-03-04                  468.0                 -2316           114.0   
b'CA' 2020-03-04                  746.0                     0           200.0   
b'CO' 2020-03-04                   56.0                   -13             NaN   
b'CT' 2020-02-29                   42.0                 -4124             NaN   
b'DC' 2020-03-05                   69.0                     0            17.0   
b'DE' 2020-03-06                   11.0                     0             4.0   
b'FL' 2020-01-29                 2005.0                     0             NaN   
b'GA' 2020-03-04                  783.0                     0             NaN   
b'GU' 2020-03-16                    0.0                     0             0.0   
b'HI' 2020-03-04                   23.0                   -14             7.0   
b'IA' 2020-03-06                   23.0                  -193            34.0   
b'ID' 2020-03-07                   15.0                    -2             3.0   
b'IL' 2020-03-04                 1326.0                     0           304.0   
b'IN' 2020-02-27                  595.0                   -13           159.0   
b'KS' 2020-03-06                  140.0                     0             1.0   
b'KY' 2020-03-06                  263.0                    -5            61.0   
b'LA' 2020-03-07                  271.0                     0             NaN   
b'MA' 2020-01-22                  290.0                   -91            47.0   
b'MD' 2020-03-05                  281.0                     0            68.0   
b'ME' 2020-03-07                    4.0                    -2             0.0   
b'MI' 2020-03-01                  471.0                     0           137.0   
b'MN' 2020-03-06                   26.0                     0            24.0   
b'MO' 2020-03-07                  412.0                     0           215.0   
b'MP' 2020-03-16                    NaN                     0             NaN   
b'MS' 2020-03-07                  465.0                   -98           113.0   
b'MT' 2020-03-07                    1.0                    -8            18.0   
b'NC' 2020-03-04                   29.0                     0           244.0   
b'ND' 2020-03-07                   10.0                    -5             5.0   
b'NE' 2020-02-15                   95.0                   -48            81.0   
b'NH' 2020-03-04                    6.0                    -7             NaN   
b'NJ' 2020-02-10                  349.0                    -6            61.0   
b'NM' 2020-03-05                   17.0                     0             NaN   
b'NV' 2020-03-05                  282.0                     0            70.0   
b'NY' 2020-03-02                  325.0                     0           109.0   
b'OH' 2020-03-05                  482.0                     0           182.0   
b'OK' 2020-03-07                   76.0                   -19            45.0   
b'OR' 2020-03-04                   91.0                   -14            27.0   
b'PA' 2020-03-03                  400.0                 -1145           301.0   
b'PR' 2020-03-16                   70.0                     0             4.0   
b'RI' 2020-03-01                   15.0                  -181             2.0   
b'SC' 2020-03-04                  387.0                     0           150.0   
b'SD' 2020-03-07                   31.0                     0            10.0   
b'TN' 2020-03-05                    9.0                     0           249.0   
b'TX' 2020-03-03                  196.0    

## Task 2 - DataFrame summarizing

Using Bloom_etal_2018_Reduced_Dataset.csv, we are going to do more dataframe manipulation and subsetting and summarizing

### Task 2.1

Read in Bloom_etal_2018_Reduced_Dataset.csv and create two new columns ('genus','species') that consists of the column taxa split at the underscore. Print out the head of this new dataframe and the number of unique genera**

hint: pd.str.split(,expand=True)

for example:

taxa	genus	species
Alosa_alabamae	Alosa	alabamae


In [122]:
bloom_df = pd.read_csv('Bloom_etal_2018_Reduced_Dataset.csv')

In [123]:
bloom_df.head()

taxa  logbodysize  trophic_position         Reg
0        Alosa_alabamae     1.707570          0.431364  diadromous
1           Alosa_alosa     1.778151          0.556303  diadromous
2          Alosa_fallax     1.778151          0.556303  diadromous
3       Alosa_mediocris     1.778151          0.612784  diadromous
4  Alosa_pseudoharengus     1.602060          0.544068  diadromous

In [125]:
bloom_df.at[49, 'taxa'] = 'Thryssa_dussumieri'

above code fixes an error in a single cell that is index number 49 in taxa column 

In [139]:
bloom_df[['Genus', 'Species']] = bloom_df['taxa'].str.split('_', expand=True)

In [140]:
bloom_df.head()

taxa  logbodysize  trophic_position         Reg  Genus  \
0        Alosa_alabamae     1.707570          0.431364  diadromous  Alosa   
1           Alosa_alosa     1.778151          0.556303  diadromous  Alosa   
2          Alosa_fallax     1.778151          0.556303  diadromous  Alosa   
3       Alosa_mediocris     1.778151          0.612784  diadromous  Alosa   
4  Alosa_pseudoharengus     1.602060          0.544068  diadromous  Alosa   

          Species  
0        alabamae  
1           alosa  
2          fallax  
3       mediocris  
4  pseudoharengus

In [142]:
bloom_df['Genus'].value_counts()

Anchoa            7
Alosa             6
Anchoviella       5
Pellonula         2
Potamalosa        1
Chirocentrodon    1
Sundasalanx       1
Hilsa             1
Cetengraulis      1
Clupeoides        1
Anodontostoma     1
Thryssa           1
Coilia            1
Microthrissa      1
Escualosa         1
Stolothrissa      1
Tenualosa         1
Amazonsprattus    1
Rhinosardinia     1
Engraulis         1
Encrasicholina    1
Ethmalosa         1
Hyperlophus       1
Gilchristella     1
Jenkinsia         1
Stolephorus       1
Clupeichthys      1
Sierrathrissa     1
Sauvagella        1
Anchovia          1
Spratelloides     1
Ehirava           1
Pellona           1
Ramnogaster       1
Name: Genus, dtype: int64

In [147]:
cnt = 0
visited = []
for i in range(0, len(bloom_df['Genus'])):
        if bloom_df['Genus'][i] not in visited:
                visited.append(bloom_df['Genus'][i])
                cnt += 1
print("Number of unique genera :",
      cnt)
print("unique values :",
     visited)

Number of unique genera : 34
unique values : ['Alosa', 'Amazonsprattus', 'Anchoa', 'Anchovia', 'Anchoviella', 'Anodontostoma', 'Cetengraulis', 'Chirocentrodon', 'Clupeichthys', 'Clupeoides', 'Coilia', 'Ehirava', 'Encrasicholina', 'Engraulis', 'Escualosa', 'Ethmalosa', 'Gilchristella', 'Hilsa', 'Hyperlophus', 'Jenkinsia', 'Microthrissa', 'Pellona', 'Pellonula', 'Potamalosa', 'Ramnogaster', 'Rhinosardinia', 'Sauvagella', 'Sierrathrissa', 'Spratelloides', 'Stolephorus', 'Stolothrissa', 'Sundasalanx', 'Tenualosa', 'Thryssa']


### Task 2.2

Create a new dataframe with the mean logbodysize and trophicposition of each genera. Sort this data frame by the largest body size. Print the head of this dataframe.

In [156]:
mean_df = bloom_df.groupby('Genus', as_index=True).agg(['mean'])

In [161]:
list(mean_df.columns)  # got an error so needed to see what the new column name was for mean logbodysize

[('logbodysize', 'mean'), ('trophic_position', 'mean')]

In [159]:
mean_df = mean_df.sort_values([('logbodysize', 'mean')], ascending = (False))

In [160]:
mean_df.head()

logbodysize trophic_position
                  mean             mean
Genus                                  
Tenualosa     1.778151         0.462398
Alosa         1.739062         0.540815
Coilia        1.544068         0.477121
Ethmalosa     1.544068         0.397940
Potamalosa    1.505150         0.518514

Which genera is the smallest and largest?

smallest is Potamaolsa and largest is Tenualosa

What is the trophic position of the smallest and largest? smallest is 0.462398 and largest is 0.518514

## Task 3 - Read in muliple files to a dictionary and make a DataFrame - OPTIONAL/BONUS

### This is not something you are expected to do in this course, but just here to give you an idea of the things that you COULD do. Answers will be posted after due date.

Using logfiles: we are going to do read in each file, get some data, append it to a dictionary to later make into a dataframe.

note: make sure to unzip logfiles